File: apps/app-frontend/src/helpers/mcarchive.ts
import { invoke } from '@tauri-apps/api/core'

export type McArchiveGameVersion = {
	id: number
	name: string
	versionType?: string | null
}

export type McArchiveFile = {
	uuid: string
	name: string
	sha256?: string | null
	archiveUrl?: string | null
	directUrl?: string | null
	redirectUrl?: string | null
	pageUrl?: string | null
}

export type McArchiveModVersion = {
	uuid: string
	name: string
	gameVersions: McArchiveGameVersion[]
	files: McArchiveFile[]
}

export type McArchiveMod = {
	uuid: string
	slug: string
	name: string
	summary?: string | null
	description?: string | null
	pageUrl?: string | null
	modVersions: McArchiveModVersion[]
}

export function getMcArchiveGameVersions() {
	return invoke<McArchiveGameVersion[]>('plugin:mcarchive|mcarchive_get_game_versions')
}

export async function searchMcArchiveMods(query: string, gameVersion?: string | null) {
	const keyword = query.trim()
	const search = (value: string, includeGameVersion = true) =>
		invoke<McArchiveMod[]>('plugin:mcarchive|mcarchive_search_mods', {
			keyword: value,
			...(includeGameVersion && gameVersion ? { gameVersion } : {}),
		})

	let results = await search(keyword)
	if (results.length === 0 && gameVersion) {
		results = await search(keyword, false)
	}
	if (results.length > 0) return results

	// MCArchive's keyword matching does not treat punctuation or whitespace as
	// interchangeable, while its project slugs commonly omit both.
	const slugLikeKeyword = keyword.replace(/[\s_-]+/g, '')
	if (slugLikeKeyword && slugLikeKeyword !== keyword) {
		const slugResults = await search(slugLikeKeyword, !!gameVersion)
		if (slugResults.length > 0) return slugResults
		if (gameVersion) {
			const unfilteredSlugResults = await search(slugLikeKeyword, false)
			if (unfilteredSlugResults.length > 0) return unfilteredSlugResults
		}
	}

	if (!keyword) return results

	// The archive API's keyword matching is intentionally conservative. Fetch the
	// small project index only when its direct query misses, then match locally.
	const catalog = await search('', !!gameVersion)
	const normalized = keyword.toLocaleLowerCase().replace(/[\s_-]+/g, '')
	return catalog.filter((project) =>
		[project.slug, project.name, project.description ?? ''].some((value) =>
			value
				.toLocaleLowerCase()
				.replace(/[\s_-]+/g, '')
				.includes(normalized),
		),
	)
}

export function getMcArchiveModBySlug(slug: string) {
	return invoke<McArchiveMod>('plugin:mcarchive|mcarchive_get_mod_by_slug', { slug })
}

File: apps/app/src/api/mcarchive.rs
use crate::api::Result;

pub fn init<R: tauri::Runtime>() -> tauri::plugin::TauriPlugin<R> {
    tauri::plugin::Builder::new("mcarchive")
        .invoke_handler(tauri::generate_handler![
            mcarchive_get_game_versions,
            mcarchive_search_mods,
            mcarchive_get_mod_by_slug,
            mcarchive_get_file_by_filename,
            mcarchive_get_file_by_sha256,
        ])
        .build()
}

#[tauri::command]
pub async fn mcarchive_get_game_versions()
-> Result<Vec<theseus::mcarchive::McArchiveGameVersion>> {
    Ok(theseus::mcarchive::get_game_versions().await?)
}

#[tauri::command]
pub async fn mcarchive_search_mods(
    keyword: &str,
    game_version: Option<&str>,
) -> Result<Vec<theseus::mcarchive::McArchiveMod>> {
    Ok(theseus::mcarchive::search_mods(keyword, game_version).await?)
}

#[tauri::command]
pub async fn mcarchive_get_mod_by_slug(
    slug: &str,
) -> Result<theseus::mcarchive::McArchiveMod> {
    Ok(theseus::mcarchive::get_mod_by_slug(slug).await?)
}

#[tauri::command]
pub async fn mcarchive_get_file_by_filename(
    filename: &str,
) -> Result<Option<theseus::mcarchive::McArchiveFile>> {
    Ok(theseus::mcarchive::get_file_by_filename(filename).await?)
}

#[tauri::command]
pub async fn mcarchive_get_file_by_sha256(
    sha256: &str,
) -> Result<Option<theseus::mcarchive::McArchiveFile>> {
    Ok(theseus::mcarchive::get_file_by_sha256(sha256).await?)
}
